functions/nodejs-apps.bash

```diff
@@ -202,3 +202,141 @@ nodered_setup() {
+    fi
+    echo -n "$(timestamp) [openHABian] Removing Zigbee2MQTT service... "
+    if ! cond_redirect systemctl stop zigbee2mqtt.service; then echo "FAILED (disable service)"; fi
+    if ! rm -f /etc/systemd/system/zigbee2mqtt.service; then echo "FAILED (remove service)"; fi
+    if cond_redirect systemctl -q daemon-reload; then echo "OK"; else  echo "FAILED (daemon-reload)"; return 1; fi
+
+    echo -n "$(timestamp) [openHABian] Uninstalling Zigbee2MQTT... "
+    if ! cond_redirect npm uninstall zigbee2mqtt ; then echo "FAILED (npm uninstall)"; fi
+    if ! rm -rf /var/log/zigbee2mqtt; then echo "FAILED (remove log)"; fi
+    if rm -rf "/opt/zigbee2mqtt"; then echo "OK"; else echo "FAILED"; fi
+    
+    if [[ -n "$INTERACTIVE" ]]; then
+      whiptail --title "Zigbee2MQTT removed" --msgbox "Zigbee2MQTT was removed from your system." 7 80
+    fi
+    return 0;
+  fi
+  if [[ $1 != "install" ]]; then return 1; fi
+
+  # if a config file exists do only update and exit
+  if [[ -e "/opt/zigbee2mqtt/data/configuration.yaml" ]] ; then
+    if [[ -n $INTERACTIVE ]]; then
+      if ! (whiptail --title "Zigbee2MQTT installation" --yes-button "Continue" --no-button "Cancel" --yesno "$z2mInstalledText" 14 80); then echo "CANCELED"; return 0; fi
+    fi
+
+    echo -n "$(timestamp) [openHABian] Updating Zigbee2MQTT... "
+    cd /opt/zigbee2mqtt || echo "FAILED (cd)"
+    if ! cond_redirect systemctl stop zigbee2mqtt ; then echo "FAILED (stop systemctl)"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" cp -R data data-backup; then echo "FAILED (cp backup)"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" git checkout HEAD -- npm-shrinkwrap.json; then echo "FAILED git"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" git pull; then echo "FAILED git"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" npm install zigbee2mqtt ; then echo "FAILED npm"; fi
+    if ! cond_redirect sudo -u "${username:-openhabian}" cp -R data-backup/* data; then echo "FAILED (cp backup)"; fi
+    rm -rf data-backup
+    cd /opt || echo "FAILED (cd)"
+    if ! cond_redirect systemctl start zigbee2mqtt; then echo "FAILED"; fi
+
+    if [[ -n $INTERACTIVE ]]; then
+      whiptail --title "Operation successful" --msgbox "$updateSuccessText" 15 80
+    fi
+    echo "OK"
+    return 0
+  fi
+
+  # get usb adapters for radio menu
+  while IFS= read -r line; do
+    my_adapters="$my_adapters $line $loopSel "
+    loopSel=0
+  done < <( ls /dev/serial/by-id )
+  unset IFS
+  
+  # ask for user input parameters
+  if [[ -n $INTERACTIVE ]]; then
+    if ! (whiptail --title "Zigbee2MQTT installation" --yes-button "Continue" --no-button "Cancel" --yesno "$introText" 14 80); then echo "CANCELED"; return 0; fi
+    if ! selectedAdapter=$(whiptail --noitem --title "Zigbee2MQTT installation" --radiolist "$adapterText" 14 100 4 $my_adapters 3>&1 1>&2 2>&3); then return 0; fi
```

Comment:
code checker is complaining because I don't quote $my_adapters. But I cannot quote because $my_adapters holds all usb adapters (we don't know how many). If I quote all arguments become one single argument. Can somebody tell me how to solve that?

Reply:
put a comment on the line before with `# shellcheck disable=<error code>` with error code being the error code the ShellCheck gives you. It should look something like `SC2086`.

Reply:
cool. It works.